cmd/ztest/ztest.c

```diff
@@ -6419,28 +6414,19 @@ ztest_run(ztest_shared_t *zs)
 	 * Kick off all the tests that run in parallel.
 	 */
 	for (t = 0; t < ztest_opts.zo_threads; t++) {
-		kthread_t *thread;
-
-		if (t < ztest_opts.zo_datasets &&
-		    ztest_dataset_open(t) != 0) {
-			umem_free(tid,
-			    ztest_opts.zo_threads * sizeof (kt_did_t));
+		if (t < ztest_opts.zo_datasets && ztest_dataset_open(t) != 0)
 			return;
-		}
 
-		VERIFY3P(thread = zk_thread_create(NULL, 0,
-		    (thread_func_t)ztest_thread,
-		    (void *)(uintptr_t)t, 0, NULL, TS_RUN, 0,
-		    PTHREAD_CREATE_JOINABLE), !=, NULL);
-		tid[t] = thread->t_tid;
+		run_threads[t] = thread_create(NULL, 0, ztest_thread,
+		    (void *)(uintptr_t)t, 0, NULL, TS_RUN, defclsyspri);
 	}
 
 	/*
 	 * Wait for all of the tests to complete.  We go in reverse order
 	 * so we don't close datasets while threads are still using them.
 	 */
 	for (t = ztest_opts.zo_threads - 1; t >= 0; t--) {
-		thread_join(tid[t]);
+		VERIFY0(pthread_join((pthread_t)run_threads[t], NULL));
```

Comment:
Good catch, I think you're right based on the `pthread_join(3)` documentation.  This is almost certainly why we created them in the detached state before.  We'll want to revert to that behavior.

```
       Failure  to  join with a thread that is joinable (i.e., one that is not
       detached), produces a "zombie thread".  Avoid doing  this,  since  each
       zombie  thread  consumes  some system resources, and when enough zombie
       threads have accumulated, it will no longer be possible to  create  new
       threads (or processes).
```